lib/Data/Validate/Sanctions/Fetcher.pm

```diff
@@ -48,62 +68,90 @@ sub _ofac_xml_zip {
+        } elsif ($dob =~ m/^(\d{4})-(\d0{1,2})-(\d{1,2})$/) {
+            $dob = $1 if $2 == 0 or $3 == 0;
+        }
+        $dob = $1 if $dob =~ m/^[A-Z][a-z]{2}-(\d{4})$/;
+
+        if ($dob =~ m/^\d{4}$/) {
+            push @year_list, $dob;
+        } elsif ($dob =~ m/(\d{4}).*to.*(\d{4})$/) {
+            push @year_list, ($1 .. $2);
+        } elsif (my $epoch = _date_to_epoch($dob)) {
+            push @epoch_list, $epoch if defined $epoch;
+        }
+    }
+
+    for my $name (@$name_list) {
+        # some names contain comma
+        $name =~ s/,//g;
+
+        $dataset->{$name}->{dob_epoch} //= [] if @epoch_list;
+        $dataset->{$name}->{dob_year}  //= [] if @year_list;
+        push @{$dataset->{$name}->{dob_epoch}}, @epoch_list;
+        push @{$dataset->{$name}->{dob_year}},  @year_list;
+        $dataset->{$name}->{dob_epoch} = [uniq $dataset->{$name}->{dob_epoch}->@*];
+        $dataset->{$name}->{dob_year}  = [uniq $dataset->{$name}->{dob_year}->@*];
+    }
 }
 
 sub _ofac_xml {
     my $content = shift;
 
-    my @names;
     my $ref = xml2hash($content, array => ['aka'])->{sdnList};
     my $ofac_ref = {};
 
     foreach my $entry (@{$ref->{sdnEntry}}) {
         next unless $entry->{sdnType} eq 'Individual';
 
-        push @names, _process_name($_->{firstName} // '', $_->{lastName} // '') for ($entry, @{$entry->{akaList}{aka} // []});
-        my $name = pop @names;
+        my @names;
 
-        $ofac_ref->{$name}->{dob_epoch} ||= [];
+        push @names, _process_name($_->{firstName} // '', $_->{lastName} // '') for ($entry, @{$entry->{akaList}{aka} // []});
 
-        my $dob = $entry->{dateOfBirthList}{dateOfBirthItem};
+        my $dobs = $entry->{dateOfBirthList}{dateOfBirthItem};
 
+        my @dob_list;
         # In one of the xml files, some of the clients have more than one date of birth
         # Hence, $dob can be either an array or a hashref
-        my @dob_list = map { $_->{dateOfBirth} || () } (ref($dob) eq 'ARRAY' ? @$dob : $dob);
-
-        foreach my $dob (@dob_list) {
-
-            # Some of the values are only years (ex. '1946')
-            # We don't want to include them
-            next unless $dob !~ /^\d{4}$/;
-
-            $dob =~ s/ /-/g;
-
-            try {
-                $dob = Date::Utility->new($dob);
-                push @{$ofac_ref->{$name}->{dob_epoch}}, $dob->epoch;
-            }
+        foreach my $dob (map { $_->{dateOfBirth} || () } (ref($dobs) eq 'ARRAY' ? @$dobs : $dobs)) {
+            push @dob_list, $dob;
         }
 
+        _process_name_and_dob(\@names, \@dob_list, $ofac_ref);
     }
 
-    die 'Datetime is invalid' unless (_validate_date($ref->{publshInformation}{Publish_Date}));
-
-    my $parser = DateTime::Format::Strptime->new(
-        pattern  => '%m/%d/%Y',
-        on_error => 'croak',
-    );
+    $ref->{publshInformation}{Publish_Date} =~ m/(\d{1,2})\/(\d{1,2})\/(\d{4})/;
+    my $publish_epoch = _date_to_epoch("$3-$1-$2");
```

Comment:
```suggestion
    my $publish_epoch = $ref->{publshInformation}{Publish_Date} =~ m/(\d{1,2})\/(\d{1,2})\/(\d{4})/
      ? _date_to_epoch("$3-$1-$2") : undef; 
```
Or similar to avoid sending undefined captures